### PR TITLE
feat(core-clp): Add `BoundedReader` to prevent out-of-bound reads in segmented input streams.

### PR DESCRIPTION
**Pull request by gibber9809**
# Description
This PR adds a BoundedReader class that can help avoid backwards seeks when reading input streams segmented into several logical chunks. The BoundedReader is a ReaderInterface that prevents reading or seeking beyond a certain "bound" byte offset.

This is used in a follow-up PR to ensure that readers for different parts of a single-file-archive being streamed over the network do not accidentally read past the end of their section (this happens frequently with readers that buffer input beyond what the user has requested such as ZstdDecompressor).

For example consider an input stream divided into the following logical chunks:
`| zstd stream 1 | header bytes | zstd stream 2|`

If a ZstdDecompressor reading zstd stream 1 directly wraps that input stream it will almost certainly end up consuming the header bytes and some parts of zstd stream 2 while populating its internal buffer. In fact this sort of speculative buffering is required if we want to  ZstdDecompressor to be performant. As a result, reading the header bytes and decompressing zstd stream 2 requires first seeking backwards in the original input stream.

BoundedReader addresses this problem by wrapping the input stream making it so that the ZstdDecompressor is unable to consume any bytes beyond the end of the logical section it belongs to, meaning that the following header and stream sections can always be read without backwards seeks.

This BoundedReader approach has some advantages over approaches that allow backwards seeking by buffering the input stream:
1) It uses less memory and requires less data-copy
2) The implementation is simple and easy to verify
3) The BoundedReader approach helps prevent a whole class of bugs involving faulty readers reading past the end of their logical section, and can help catch issues with corrupt archives

# Validation performed
* Added tests for seeking and reading edge cases

## Summary by CodeRabbit

- **New Features**
	- Introduced the `BoundedReader` class to manage reading limits within input streams.
	- Added unit tests for `BoundedReader` functionalities, ensuring robust error handling and boundary checks.

- **Bug Fixes**
	- Enhanced error handling in the `StringReader` class to prevent out-of-bounds seeking.

### CODE REVIEW

**Review by haiqi96 — COMMENTED** [review]
Talked with Devin offline and there is a concrete example that explains the usecase of this class. Devin, please add it to the PR when you get a chance.

In general the PR makes sense to me. While there might be a more elegant design to achieve the same target, I feel we can go with the design in this PR given 
1.the class design is simple and straight forward 
2. We have a rather tight deadline , including other upcoming changes.

Left a few comments, and also we can rename the class to be BoundedReader. Unless @kirkrodrigues has other naming suggestions

Note this class has different purpose from BufferedFileReader. The BufferedFileReader is designed for user who knows a specific pos they will need to seek back, but this class is designed for user who knows a specific pos that they don't want to seek beyond.

File: components/core/src/clp/StringReader.cpp [hunk 1]
```diff
@@ -41,6 +41,10 @@ ErrorCode StringReader::try_read(char* buf, size_t num_bytes_to_read, size_t& nu
 }
 
 ErrorCode StringReader::try_seek_from_begin(size_t pos) {
+    if (pos > input_string.size()) {
```
[the remark this answers is not in the document]
Comment (gibber9809) [reply components/core/src/clp/StringReader.cpp:44]: I would classify this as a bug I'm fixing in the StringReader class so that I can use it for tests. Every other reader we have will EOF if you seek past the end from what I've seen.

Actually, the way the rest of this class is written if you first seek past the end of the input buffer it will happily let you read data beyond the end of the buffer. I.e. it is very explicitly a bug.

File: components/core/tests/test-CheckpointReader.cpp [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        REQUIRE(clp::ErrorCode_Success == rc);
+        REQUIRE(1 == num_bytes_read);
+        REQUIRE(1 == string_reader.get_pos());
+        REQUIRE(1 == checkpoint_reader.get_pos());
+    }
+
+    SECTION("CheckpointReader does allow reads before checkpoint.") {
+        clp::StringReader string_reader;
+        string_reader.open(cTestString);
+        clp::CheckpointReader checkpoint_reader{&string_reader, 1};
```
[the remark this answers is not in the document]
Comment (gibber9809) [reply components/core/tests/test-CheckpointReader.cpp:51]: No, "1" is correct. This test confirms that the checkpoint reader allows reading byte 0 which is before byte 1.

File: components/core/src/clp/StringReader.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (haiqi96) [reply components/core/src/clp/StringReader.cpp:44]: Hmm, The FileReader internally calls fseeko, which I believe would allow seeking beyond the end of file.

BufferedFileReader and BufferReader would return ErrorCode_Truncated, and won't update the pos at all if the pos is greater than the maximum length,

From the consistency point of view, maybe we should let it return ErrorCode_Truncated. But also wonder if the rest of your code is dependent on the current behavior.

File: components/core/tests/test-CheckpointReader.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (haiqi96) [reply components/core/tests/test-CheckpointReader.cpp:56]: oh I see.

File: components/core/src/clp/StringReader.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply components/core/src/clp/StringReader.cpp:44]: * The implementations of ReaderInterface have inconsistent behaviour.
  * Some (e.g., `FileReader`) will rely on the lower implementation to return an error if the seek fails.
  * Some (e.g. `BufferReader`) will return an error if the seek is past the end, but they won't modify `m_pos`.
  * Some (e.g., `NetworkReader`) will return an error if the seek is past the end, and they will modify `m_pos`.
  * Each of the above returns different error codes.

I think the practical implementation is for `try_seek_from_begin` to:
* try to seek until `pos`
* if that's past the end of the medium (file/buffer/etc.), `m_pos` should be updated to just past the last byte.
* the method should return `ErrorCode_EndOfFile`.

The reason to update the `pos` even though we don't get 

If y'all agree, we should open a GH issue to refactor the existing implementations. And for this implementation, we implement the proposal above (basically what Devin's implemented---although on error, we're still updating `m_cur_pos`, even tough the error may not be EOF?).

**Comment by kirkrodrigues** [comment]
I'm fine with `BoundedReader`.

File: components/core/src/clp/StringReader.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (haiqi96) [reply components/core/src/clp/StringReader.cpp:44]: Sure, the standard behavior sounds good to me.

[the remark this answers is not in the document]
Comment (gibber9809) [reply components/core/src/clp/StringReader.cpp:44]: That makes sense to me.

Also went and updated BoundedReader to only update `m_curr_pos` on error if that error is EOF.

File: components/core/src/clp/BoundedReader.cpp [hunk 3]
```diff
@@ -0,0 +1,39 @@
+#include "BoundedReader.hpp"
+
+namespace clp {
+auto BoundedReader::try_seek_from_begin(size_t pos) -> ErrorCode {
+    auto next_pos = pos > m_bound ? m_bound : pos;
```
Comment (haiqi96) [inline components/core/src/clp/BoundedReader.cpp:5]: nit:
```suggestion
    auto const next_pos = pos > m_bound ? m_bound : pos;
```

Comment (haiqi96) [inline components/core/src/clp/BoundedReader.cpp:5]: nit: if I am not mistaken, this can be a const
```suggestion
    auto const next_pos = pos > m_bound ? m_bound : pos;
```

File: components/core/src/clp/BoundedReader.cpp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    if (auto const rc = m_reader->try_seek_from_begin(next_pos); ErrorCode_Success != rc) {
+        m_curr_pos = ErrorCode_EndOfFile == rc ? next_pos : m_curr_pos;
+        return rc;
+    }
+    m_curr_pos = next_pos;
+    if (m_curr_pos >= m_bound) {
+        return ErrorCode_EndOfFile;
+    }
+    return ErrorCode_Success;
```
Comment (haiqi96) [inline components/core/src/clp/BoundedReader.cpp:14]: Just thinking out loud, in an ideal world, `rc = m_reader->try_seek_from_begin(next_pos)` should guarantee that `m_curr_pos >= m_bound`. So the code can be rewritten as the follow.

It doesn't really simplify the code a lot though, so it's totally up to you to choose which version to use.

```suggestion
    if (auto const rc = m_reader->try_seek_from_begin(next_pos); 
        ErrorCode_Success != rc && ErrorCode_EndOfFile != rc) {
        return rc;
    }
    m_curr_pos = next_pos;
    if (m_curr_pos >= m_bound) {
        return ErrorCode_EndOfFile;
    }
    return ErrorCode_Success;
```

**Comment by haiqi96** [comment]
Left another two comments for nit. otherwise the code looks good to me

File: components/core/src/clp/BoundedReader.cpp (hunk 4 above)
Comment (gibber9809) [reply components/core/src/clp/BoundedReader.cpp:14]: I think this one I'll leave because I have to think about this version of the code for an extra second before deciding that it works.

**Review by haiqi96 — DISMISSED** [review]
Looks good to me. @kirkrodrigues do you want to do another round of review?

**Comment by haiqi96** [comment]
@gibber9809 Can you please also open an issue to track the proposed change for commonizing seek interface?

**Comment by gibber9809** [comment]
> @gibber9809 Can you please also open an issue to track the proposed change for commonizing seek interface?

Added issue #628.

File: components/core/src/clp/BoundedReader.hpp [hunk 5]
```diff
@@ -0,0 +1,78 @@
+#ifndef CLP_BOUNDEDREADER_HPP
+#define CLP_BOUNDEDREADER_HPP
+
+#include "ReaderInterface.hpp"
```
Comment (LinZhihao-723) [inline components/core/src/clp/BoundedReader.hpp:4]: We are missing some headers includes for:
- ErrorCode
- size_t

File: components/core/src/clp/BoundedReader.hpp [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+     * @param num_bytes_to_read The number of bytes to try and read
+     * @param num_bytes_read The actual number of bytes read
+     * @return ErrorCode_errno on error
+     * @return ErrorCode_EndOfFile on EOF or trying to read after hitting checkpoint
+     * @return ErrorCode_Success on success
+     */
+    auto
+    try_read(char* buf, size_t num_bytes_to_read, size_t& num_bytes_read) -> ErrorCode override;
+
+    auto try_read_to_delimiter(char delim, bool keep_delimiter, bool append, std::string& str)
```
Comment (LinZhihao-723) [inline components/core/src/clp/BoundedReader.hpp:66]: Shall we add a doc string to explain why we override the default implementation?

Comment (LinZhihao-723) [inline components/core/src/clp/BoundedReader.hpp:66]: Since parameters are not used, shall we add `[[maybe_unused]]` to silence clang-tidy warnings?

File: components/core/src/clp/BoundedReader.hpp [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    // Methods implementing the ReaderInterface
+    /**
+     * Tries to get the current position of the read head in the underlying reader.
+     * @param pos Position of the read head in the underlying reader
+     * @return ErrorCode_Success on success
+     * @return ErrorCode_errno on failure
+     */
+    auto try_get_pos(size_t& pos) -> ErrorCode override { return m_reader->try_get_pos(pos); }
```
Comment (LinZhihao-723) [inline components/core/src/clp/BoundedReader.hpp:38]: We should add `[[nodiscard]]` for any non-void returns

File: components/core/src/clp/BoundedReader.hpp [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    auto
+    try_read(char* buf, size_t num_bytes_to_read, size_t& num_bytes_read) -> ErrorCode override;
+
+    auto try_read_to_delimiter(char delim, bool keep_delimiter, bool append, std::string& str)
+            -> ErrorCode override {
+        return ErrorCode_Unsupported;
+    }
+
+private:
+    ReaderInterface* m_reader{nullptr};
```
Comment (LinZhihao-723) [inline components/core/src/clp/BoundedReader.hpp:72]: Using a pointer here doesn't enforce the given reader interface to have a lifetime >= the bounded reader. To be safe, shall we consider using a shared pointer instead?
What do u think? @kirkrodrigues

File: components/core/src/clp/BoundedReader.hpp [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        m_curr_pos = m_reader->get_pos();
+        if (m_curr_pos > m_bound) {
+            throw ReaderInterface::OperationFailed(ErrorCode_BadParam, __FILE__, __LINE__);
+        }
+    }
+
+    // Methods implementing the ReaderInterface
+    /**
+     * Tries to get the current position of the read head in the underlying reader.
+     * @param pos Position of the read head in the underlying reader
```
Comment (LinZhihao-723) [inline components/core/src/clp/BoundedReader.hpp:34]: This is an output parameter, so it should start with `Returns xxx`
```suggestion
     * @param pos Returns the position of the underlying reader's head
```

File: components/core/src/clp/BoundedReader.hpp [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+ * want to prevent a reader for an earlier segment consuming any bytes from a later segment. In
+ * particular, reading part of a later segment may force the reader for that later segment to seek
+ * backwards, which can be either inefficient or impossible for certain kinds of input streams.
+ */
+class BoundedReader : public ReaderInterface {
+public:
+    // Constructor
+    explicit BoundedReader(ReaderInterface* reader, size_t bound)
+            : m_reader(reader),
+              m_bound(bound) {
```
Comment (LinZhihao-723) [inline components/core/src/clp/BoundedReader.hpp:21]:
```suggestion
            : m_reader{reader},
              m_bound{bound} {
```
We should enforce `{}` initializations in constructors' initialization list

File: components/core/src/clp/BoundedReader.hpp [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+     * @param pos Position of the read head in the underlying reader
+     * @return ErrorCode_Success on success
+     * @return ErrorCode_errno on failure
+     */
+    auto try_get_pos(size_t& pos) -> ErrorCode override { return m_reader->try_get_pos(pos); }
+
+    /**
+     * Tries to seek to the given position, limited by the checkpoint. If pos is past the checkpoint
+     * then this reader seeks up to the checkpoint in the underlying stream, then returns
+     * ErrorCode_EndOfFile on success or ErrorCode_errno otherwise.
```
Comment (LinZhihao-723) [inline components/core/src/clp/BoundedReader.hpp:43]: This description is duplicates the return section. I think we can drop the failure case.
```suggestion
     * Tries to seek to the given position, limited by the checkpoint.
```

File: components/core/src/clp/BoundedReader.hpp [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+     * @return ErrorCode_EndOfFile on EOF or if trying to seek beyond the checkpoint
+     * @return ErrorCode_errno on failure
+     */
+    auto try_seek_from_begin(size_t pos) -> ErrorCode override;
+
+    /**
+     * Tries to read up to a given number of bytes from the file, limited by the checkpoint. If the
+     * requested read would advance the read head beyond the checkpoint then the read size is
+     * adjusted such that the read head is advanced to the checkpoint. If a read is attempted when
+     * the read head is at the checkpoint then ErrorCode_EndOfFile is returned.
```
Comment (LinZhihao-723) [inline components/core/src/clp/BoundedReader.hpp:55]: Similar to the previous comment, it has duplicated descriptions in the return section.
```suggestion
     * Tries to read up to a given number of bytes from the file, limited by the checkpoint.
```
Also, I think if the read request advances the head beyond the checkpoint, we should invalidate the read request and set it to EOF, right? The current implementation just swallows the error instead

File: components/core/src/clp/BoundedReader.cpp [hunk 13]
```diff
@@ -0,0 +1,39 @@
+#include "BoundedReader.hpp"
```
Comment (LinZhihao-723) [inline components/core/src/clp/BoundedReader.cpp:1]: Same as the header; please ensure all headers used are included

File: components/core/src/clp/BoundedReader.cpp [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+auto BoundedReader::try_read(char* buf, size_t num_bytes_to_read, size_t& num_bytes_read)
+        -> ErrorCode {
+    if (m_curr_pos == m_bound) {
+        num_bytes_read = 0;
+        return ErrorCode_EndOfFile;
+    }
+
+    if ((m_curr_pos + num_bytes_to_read) > m_bound) {
+        num_bytes_to_read = m_bound - m_curr_pos;
+    }
```
Comment (LinZhihao-723) [inline components/core/src/clp/BoundedReader.cpp:26]: As mentioned in another comment, we shouldn't swallow this error

**Review by LinZhihao-723 — CHANGES_REQUESTED** [review]
Reviewed the header + implementation files, didn't go through unit tests yet

File: components/core/src/clp/BoundedReader.hpp (hunk 8 above)
Comment (gibber9809) [reply components/core/src/clp/BoundedReader.hpp:72]: I understand where you're coming from, but I decided not to use shared_ptr here for a few reasons.

1. The lifetime invariant is fairly simple to maintain in context, and any code that's using BoundedReader for its intended use will maintain the invariant automatically.
Specifically BoundedReader acts as a way to safely delegate ranges of an input stream to a series of different readers, which each relinquish (usually explicitly) the BoundedReader once they're finished with their section; the owner has to hold onto the underlying reader through all of these delegations so that it can correctly close the input stream at the end.

You would have to use BoundedReader in a somewhat contrived way outside of its intended use-case to violate the lifetime invariant, and anyone who does so will likely notice because their code will crash.

2. Forcing shared_ptr pollutes shared_ptr throughout the codebase even for very simple uses.

3. Forcing shared_ptr adds meaningful overhead in contexts where you may want to delegate thousands of small ranges (e.g. when reading tables from an archive).

4. Directly passing a pointer to another reader follows our existing convention for wrapping readers (e.g. our decompression classes taking raw FileReader*)

File: components/core/src/clp/BoundedReader.cpp (hunk 14 above)
Comment (gibber9809) [reply components/core/src/clp/BoundedReader.cpp:26]: This is not an error.

File: components/core/src/clp/BoundedReader.hpp (hunk 12 above)
Comment (gibber9809) [reply components/core/src/clp/BoundedReader.hpp:55]: Again, not an error. Our readers typically only present EOF once they EOF and the read size is zero.

File: components/core/src/clp/BoundedReader.hpp (hunk 6 above)
Comment (gibber9809) [reply components/core/src/clp/BoundedReader.hpp:66]: Reason is really just that BoundedReader can't delegate to a potentially more efficient implementation in the underlying reader (since it won't respect the bounds), and most code really shouldn't use this interface anyway since its a performance trap.

Can add a docstring saying as much.

File: components/core/src/clp/BoundedReader.hpp (hunk 5 above)
Comment (gibber9809) [reply components/core/src/clp/BoundedReader.hpp:4]: I know that clang-tidy cares, but isn't the official policy that we don't have to include headers that will definitely be included by the headers that we do include.

If we always defer to clang-tidy then we should change our coding guidelines.

File: components/core/src/clp/BoundedReader.cpp (hunk 13 above)
Comment (gibber9809) [reply components/core/src/clp/BoundedReader.cpp:1]: See other comment.

File: components/core/src/clp/BoundedReader.hpp [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+    // Constructor
+    explicit BoundedReader(ReaderInterface* reader, size_t bound)
+            : m_reader{reader},
+              m_bound{bound} {
+        if (nullptr == m_reader) {
+            throw ReaderInterface::OperationFailed(ErrorCode_BadParam, __FILE__, __LINE__);
+        }
+        m_curr_pos = m_reader->get_pos();
+        if (m_curr_pos > m_bound) {
+            throw ReaderInterface::OperationFailed(ErrorCode_BadParam, __FILE__, __LINE__);
```
[the remark this answers is not in the document]
Comment (gibber9809) [reply components/core/src/clp/BoundedReader.hpp:27]: This is fine. The constructor throws on every other error, and this is more of the same.

File: components/core/src/clp/BoundedReader.hpp (hunk 12 above)
Comment (LinZhihao-723) [reply components/core/src/clp/BoundedReader.hpp]: Yeah, I double-checked that this should be the expected behavior. My bad.

File: components/core/src/clp/BoundedReader.hpp (hunk 6 above)
Comment (LinZhihao-723) [reply components/core/src/clp/BoundedReader.hpp]: Sure, let's add a doc string

File: components/core/src/clp/BoundedReader.hpp (hunk 5 above)
Comment (LinZhihao-723) [reply components/core/src/clp/BoundedReader.hpp:4]: - Sorry for the confusion. Can u point me which policy u referred to? Our latest guideline is we should directly include all the headers needed (same as google: https://google.github.io/styleguide/cppguide.html#Include_What_You_Use). I should update this guideline if it's still out-dated.
- I know that clang-tidy is not officially turned on in clp core, but for any newly added files we should try to fix all clang-tidy issues. Kirk and I were trying to proceed clang-tidy in our core, and hopefully we will have it in our automated workflow soon. It makes our life easier if we start following the clang-tidy rules

Comment (gibber9809) [reply components/core/src/clp/BoundedReader.hpp:4]: See related section of our style guide https://gitlab.dev.yscope.io/clp/CLP/-/wikis/CPP-style-guide#names-and-order-of-includes

Comment (gibber9809) [reply components/core/src/clp/BoundedReader.hpp:4]: I interpret the "unless foo.hpp explicitly demonstrates its intent to provide you the symbols of bar.hpp" clause to cover situations where foo.hpp includes those symbols in the interface that it provides.

Comment (gibber9809) [reply components/core/src/clp/BoundedReader.hpp:4]: I think the current style guide is better than strict include what you use on the basis that including headers that are guaranteed to be provided by another header you include wastes vertical space and decreases readability. I would object to changing the style guide in this case.

Comment (LinZhihao-723) [reply components/core/src/clp/BoundedReader.hpp:4]: - The guidelines on Gitlab are out of date and we will update them, as well as the notion page
- We want to rely on linters for most things so that we can automate code checks as much as possible. That means we should just stick to one rule that the linter checks
- Adding a few headers recommended by clang-tidy doesn't seem like a huge investment. For now, we have to manually pay the cost of running clang-tidy, but we're working to address that. Also if u're using CLion, clang-tidy should be automatically enabled in the editor to show all warnings

Comment (gibber9809) [reply components/core/src/clp/BoundedReader.hpp:4]: I agree in with what you're saying and I will make this change, but please:
1) Keep the style guide up to date
2) Announce changes to the style guide, preferably preceded by RFC

Otherwise what you're asking me to do is strictly follow any style rules you decide on in the moment, which is an absurd coding standard.

**Review by LinZhihao-723 — CHANGES_REQUESTED** [review]
The implementation looks good to me. Sorry that there are still a few suggestions for fixing clang-tidy warnings in the test file. Your IDE/clang-tidy run might also complain about `REQUIRE` macro, but we can ignore that for now since we're planning to fix this by upgrading catch2 to a higher version.

**Comment by gibber9809** [comment]
> The implementation looks good to me. Sorry that there are still a few suggestions for fixing clang-tidy warnings in the test file. Your IDE/clang-tidy run might also complain about `REQUIRE` macro, but we can ignore that for now since we're planning to fix this by upgrading catch2 to a higher version.

Interesting. Yeah I didn't get any clang-tidy warnings (besides some incorrect ones) running from the command line likely because it was getting confused about the catch2 macro expansions. I'm guessing clion quietly does a lot of extra configuration to deal with this sort of thing.

File: components/core/tests/test-BoundedReader.cpp [hunk 16]
```diff
@@ -0,0 +1,99 @@
+#include <string>
+
+#include <Catch2/single_include/catch2/catch.hpp>
+
+#include "../src/clp/BoundedReader.hpp"
+#include "../src/clp/ErrorCode.hpp"
+#include "../src/clp/StringReader.hpp"
+
+TEST_CASE("Test Bounded Reader", "[BoundedReader]") {
+    constexpr char cTestString[]{"0123456789"};
```
[the remark this answers is not in the document]
Comment (gibber9809) [reply components/core/tests/test-BoundedReader.cpp:10]: I'm using constexpr char[] because StringReader takes std::string const& and I don't want to manually initialize an std::string every time I create a StringReader.

[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/tests/test-BoundedReader.cpp:12]: - This was to follow our guideline here: https://www.notion.so/yscope/WIP-Coding-Guidelines-9a308b847a5343958ba3cb97a850be66?pvs=4#13604e4d9e6b80b09a18d4f71a89f1c8
- From compiler's perspective using `constexpr char[]` and `constexpr std::string_view` won't make a difference: check [here](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:2,endLineNumber:19,positionColumn:2,positionLineNumber:19,selectionStartColumn:2,selectionStartLineNumber:19,startColumn:2,startLineNumber:19),source:'//+Online+C%2B%2B+compiler+to+run+C%2B%2B+program+online%0A%23include+%3Ciostream%3E%0A%23include+%3Cstring%3E%0A%23include+%3Cstring_view%3E%0A%23include+%3Ccstddef%3E%0A%0Asize_t+foo(std::string+const%26+s)+%7B%0A++++return+s.size()%3B%0A%7D%0A%0Aint+main()+%7B++++%0A++++constexpr+std::string_view+sv%7B%22Hi%22%7D%3B%0A++++constexpr+char+cstr%5B%5D%7B%22Hi%22%7D%3B%0A++++%0A++++size_t+const+a%7Bfoo(std::string%7Bsv%7D)%7D%3B%0A++++size_t+const+b%7Bfoo(cstr)%7D%3B%0A%0A++++return+a+-+b%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:g142,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'-O2',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+gcc+14.2+(Editor+%231)',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

**Review by LinZhihao-723 — APPROVED** [review]
For the PR title, how about:
```
feat(core-clp): Add `Boun`
